### PR TITLE
do not allow multiple uses of the same key

### LINKED ISSUE #59: token can be used several times
1. login using a token
2. remember the token
3. logout
4. login again using the same token

From [RFC6238](https://tools.ietf.org/html/rfc6238) :
```
   Note that a prover may send the same OTP inside a given time-step
   window multiple times to a verifier.  The verifier MUST NOT accept
   the second attempt of the OTP after the successful validation has
   been issued for the first OTP, which ensures one-time only use of an
   OTP.
```

### PR DESCRIPTION
**Pull request by karakayasemi**
resolves #59 . 
I tested it in my local. I would like to add unit test for Totp service, but, IMHO it needs some bigger refactoring. There is an object that is violating dependency injection.

### CODE REVIEW

**Comment by individual-it** [comment]
will this only store the last token or multiple of the past?

**Comment by individual-it** [comment]
tested. this fixes the issue, by saving the last token, still its possible to use the pre-last token to login again (obviously within the time limits)
Is that a problem?

**Comment by karakayasemi** [comment]
Hmm, we are also validating before and next token. https://github.com/owncloud/twofactor_totp/blob/master/lib/Service/Totp.php#L115 I thought only last used token is enough. Actually, since we are implementing this for the case of violation of user-server communication privacy,  preventing consecutive usage of the same token is enough.

**Comment by individual-it** [comment]
@PVince81 @DeepDiver1975 if you have a minute during the conference you maybe could have a look here

File: lib/Db/TotpSecret.php [hunk 1]
```diff
@@ -39,6 +42,9 @@ class TotpSecret extends Entity {
     /** @var string */
     protected $secret;
 
+	/** @var string */
+	protected $lastValidatedKey;
```
Comment (PVince81) [inline lib/Db/TotpSecret.php]: indent

File: lib/Service/Totp.php [hunk 2]
```diff
@@ -109,10 +109,18 @@ public function validateSecret(IUser $user, $key) {
...
 
-        $secret = $this->crypto->decrypt($dbSecret->getSecret());
-
-        $otp = new Otp();
-        return $otp->checkTotp(Base32::decode($secret), $key, 3);
+        if ($dbSecret->getLastValidatedKey() !== $key) {
+			$secret = $this->crypto->decrypt($dbSecret->getSecret());
+			$otp = new Otp();
+
+			if ($otp->checkTotp(Base32::decode($secret), $key, 3) === true) {
```
Comment (PVince81) [inline lib/Service/Totp.php]: add some PHPDoc comment, it is not clear what and why we are doing this

**Comment by PVince81** [comment]
> I tested it in my local. I would like to add unit test for Totp service

if really not possible, please make a technical debt ticket and explain what needs to be refactored

is this at least covered by acceptance tests ? @individual-it

**Comment by patrickjahns** [comment]
@PVince81 
No - there are no acceptance tests for this app - this app has always been tested manually

**Comment by PVince81** [comment]
5. release the current app without this PR, so we get the important fix and @karakayasemi has more time to do this properly

thoughts ?

**Comment by karakayasemi** [comment]
@PVince81 let me refactor it, I will make it today, after work hours.

**Comment by karakayasemi** [comment]
I converted spaces to tabs in the files to follow ownCloud code style, now it became difficult to review. I am closing this PR, and creating new PR with the app's code style. We can fix cs mistakes by introducing php-cs-fixer later. Let's continue in here: #63

**Pull request closed without merging**